ultralytics/engine/results.py

```diff
@@ -327,6 +334,100 @@ def verbose(self):
+                        n: {
+                            "label": self.names.get(lbl_idx[n]),
+                            "conf": float(all_conf[n]),
+                            "id": int(all_ids[n]) if all_ids is not None else None,
+                            "xyxy": r.boxes.xyxy[n] if box else None,
+                            "nxyxy": r.boxes.xyxyn[n] if box else None,
+                            "xywh": r.boxes.xywh[n] if box else None,
+                            "nxywh": r.boxes.xywhn[n] if box else None,
+                            "mask-xy": self.masks.xy[n] if mask else None,  # NOTE cpu + numpy missing attribute
+                            "mask-xyn": self.masks.xyn[n] if mask else None,  # NOTE cpu + numpy missing attribute
+                            "kp-xy": r.keypoints.xy[n] if kp else None,
+                            "kp-xyn": r.keypoints.xyn[n] if kp else None,
+                            "kp-conf": r.keypoints.conf[n] if kp else None,
+                            "xywhr": r.obb.xywhr[n] if obb else None,
+                            "xyxyxyxy": r.obb.xyxyxyxy[n] if obb else None,
+                            "xyxyxyxyn": r.obb.xyxyxyxyn[n] if obb else None,
+                            "top1": r.probs.top1 if probs else None,
+                            "top1conf": r.probs.top1conf if probs else None,
+                            "top5": r.probs.top5 if probs else None,
+                            "top5conf": r.probs.top5conf if probs else None,
+                        }
+                        for n in range(len(lbl_idx))
+                    }
+                }
+            }
+        self.result_dict = ops.to_py_types(self.result_dict) if py_typed else self.result_dict
+        return self.result_dict
+
+    def to_pandas(self, df_in: DataFrame = None) -> DataFrame:
+        """
+        Converts the results to a pandas DataFrame.
+
+        Args:
+            df_in (DataFrame, optional): Input DataFrame to append results to; defaults to None.
+
+        Returns:
+            DataFrame: Pandas Multi-Index DataFrame containing the results.
+
+        Example:
+            ```python
+            from ultralytics import YOLO
+
+            model = YOLO('yolov8s.pt')
+            results = model.predict(['bus.jpg', 'zidane.jpg'])
+
+            df = None # initialize variable
+            for result in results:
+                df = result.to_pandas(df)
+
+            df.dropna(axis=1) # View with NaN value columns dropped
+
+            # Use DataFrame methods to save results to CSV, Excel, JSON, etc.
+            ```
+        """
+        df_out = df_in if df_in is not None and isinstance(df_in, DataFrame) else DataFrame()
+        od = self.asdict()
+        k = next(iter(od.keys()))
+        # Construct DataFrame with Multi-Index
+        cols = list(od.get(k).get("detections").get(0).keys())
+        idx = MultiIndex.from_product(
+            [[k], [*list(od.get(k).get("detections").keys())]],
```

Comment:
This fails when there is no detection for a frame/image. Probably we could have the column names of the DF predefined

Reply:
Another good catch. I wrote a short loop to test scenarios both with no detections and detections, this should now be good for all tasks. The resulting `DataFrame` will output with `None` values and all `dict` keys from `asdict()` method when there are no detections.

```
                      label  conf    id  xyxy nxyxy  xywh nxywh mask-xy mask-xyn kp-xy kp-xyn kp-conf xywhr xyxyxyxy xyxyxyxyn  top1 top1conf  top5 top5conf
image      detections
image0.jpg 0           None  None  None  None  None  None  None    None     None  None   None    None  None     None      None  None     None  None     None
```